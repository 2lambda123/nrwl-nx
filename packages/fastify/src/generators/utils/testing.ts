import { Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';

export function createTreeWithFastifyApplication(appName: string): Tree {
  const tree = createTreeWithEmptyWorkspace();
  tree.write(
    'workspace.json',
    String.raw`
      {
        "projects": {
          "${appName}": {
            "root": "apps/api",
            "sourceRoot": "apps/api/src",
            "projectType": "application",
            "prefix": "api",
            "targets":{}
          }
        }
      }
    `
  );

  return tree;
}
