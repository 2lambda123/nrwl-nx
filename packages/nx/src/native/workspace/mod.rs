mod get_nx_workspace_files;
mod types;
