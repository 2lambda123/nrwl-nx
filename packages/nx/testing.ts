export {cold, hot} from 'jasmine-marbles';
export {readAll} from './src/testing-utils';
