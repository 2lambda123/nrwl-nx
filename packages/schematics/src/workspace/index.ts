import {apply, branchAndMerge, chain, externalSchematic, mergeWith, move, Rule, template, Tree, url, schematic} from '@angular-devkit/schematics';
import {Schema} from './schema';
import {names, toFileName} from '@nrwl/schematics';
import * as path from 'path';
import {ngrxVersion, nxVersion, schematicsVersion} from '../utility/lib-versions';
import * as fs from 'fs';
import {join} from 'path';
import {updateJsonFile} from '../utility/fileutils';

function updatePackageJson() {
  return (host: Tree) => {
    if (!host.exists('package.json')) {
      throw new Error('Cannot find package.json');
    }
    const packageJson = JSON.parse(host.read('package.json')!.toString('utf-8'));
    if (!packageJson.devDependencies) {
      packageJson.devDependencies = {};
    }
    if (!packageJson.dependencies) {
      packageJson.dependencies = {};
    }
    if (!packageJson.dependencies['@nrwl/nx']) {
      packageJson.dependencies['@nrwl/nx'] = nxVersion;
    }
    if (!packageJson.dependencies['@ngrx/store']) {
      packageJson.dependencies['@ngrx/store'] = ngrxVersion;
    }
    if (!packageJson.dependencies['@ngrx/router-store']) {
      packageJson.dependencies['@ngrx/router-store'] = ngrxVersion;
    }
    if (!packageJson.dependencies['@ngrx/effects']) {
      packageJson.dependencies['@ngrx/effects'] = ngrxVersion;
    }
    if (!packageJson.dependencies['@ngrx/store-devtools']) {
      packageJson.dependencies['@ngrx/store-devtools'] = ngrxVersion;
    }
    if (!packageJson.devDependencies['@nrwl/schematics']) {
      packageJson.devDependencies['@nrwl/schematics'] = schematicsVersion;
    }
    host.overwrite('package.json', JSON.stringify(packageJson, null, 2));
    return host;
  };
}

function updateAngularCLIJson(options: Schema) {
  return (host: Tree) => {
    if (!host.exists('.angular-cli.json')) {
      throw new Error('Cannot find .angular-cli.json');
    }
    const angularCliJson = JSON.parse(host.read('.angular-cli.json')!.toString('utf-8'));
    if (angularCliJson.apps.length !== 1) {
      throw new Error('Can only convert projects with one app');
    }

    angularCliJson.lint =
        [{'project': './tsconfig.app.json'}, {'project': './tsconfig.spec.json'}, {'project': './tsconfig.e2e.json'}];

    const app = angularCliJson.apps[0];
    app.root = path.join('apps', options.name, app.root);
    app.outDir = path.join('dist', 'apps', options.name);
    app.test = '../../../test.js';
    app.tsconfig = '../../../tsconfig.app.json';
    app.testTsconfig = '../../../tsconfig.spec.json';
    app.scripts = app.scripts.map((p) => path.join('../../', p));
    if (!angularCliJson.defaults) {
      angularCliJson.defaults = {};
    }
    if (!angularCliJson.defaults.schematics) {
      angularCliJson.defaults.schematics = {};
    }
    angularCliJson.defaults.schematics['collection'] = '@nrwl/schematics';
    angularCliJson.defaults.schematics['newProject'] = ['app', 'lib'];

    host.overwrite('.angular-cli.json', JSON.stringify(angularCliJson, null, 2));

    return host;
  };
}

function updateTsConfigsJson(options: Schema) {
  return (host: Tree) => {
    const angularCliJson = JSON.parse(host.read('.angular-cli.json')!.toString('utf-8'));
    const npmScope = options && options.npmScope ? options.npmScope : options.name;

    updateJsonFile('tsconfig.json', (json) => setUpCompilerOptions(json, npmScope));

    updateJsonFile('tsconfig.app.json', (json) => {
      json['extends'] = './tsconfig.json';
      if (!json.exclude) json.exclude = [];
      json.exclude = dedup(json.exclude.concat(['**/*.spec.ts', '**/*.e2e-spec.ts', 'node_modules', 'tmp']));
      setUpCompilerOptions(json, npmScope);
    });

    updateJsonFile('tsconfig.spec.json', (json) => {
      json['extends'] = './tsconfig.json';
      if (!json.exclude) json.exclude = [];
      json.files = ['test.js'];
      json.exclude = dedup(json.exclude.concat(['node_modules', 'tmp']));
      setUpCompilerOptions(json, npmScope);
    });

    updateJsonFile('tsconfig.e2e.json', (json) => {
      json['extends'] = './tsconfig.json';
      if (!json.exclude) json.exclude = [];
      json.exclude = dedup(json.exclude.concat(['**/*.spec.ts', 'node_modules', 'tmp']));
      setUpCompilerOptions(json, npmScope);
    });

    return host;
  };
}

function updateProtractorConf() {
  return (host: Tree) => {
    if (!host.exists('protractor.conf.js')) {
      throw new Error('Cannot find protractor.conf.js');
    }
    const protractorConf = host.read('protractor.conf.js')!.toString('utf-8');
    const updatedConf = protractorConf.replace(`./e2e/**/*.e2e-spec.ts`, `./apps/**/*.e2e-spec.ts`)
                            .replace(`e2e/tsconfig.e2e.json`, `./tsconfig.e2e.json`);

    host.overwrite('protractor.conf.js', updatedConf);

    return host;
  };
}

function setUpCompilerOptions(tsconfig: any, npmScope: string): void {
  if (!tsconfig.compilerOptions.paths) {
    tsconfig.compilerOptions.paths = {};
  }
  tsconfig.compilerOptions.baseUrl = '.';
  tsconfig.compilerOptions.paths[`@${npmScope}/*`] = ['libs/*'];
}

function moveFiles(options: Schema) {
  return (host: Tree) => {
    const angularCliJson = JSON.parse(host.read('.angular-cli.json')!.toString('utf-8'));
    const app = angularCliJson.apps[0];

    fs.mkdirSync('apps');
    fs.mkdirSync('libs');
    fs.unlinkSync(path.join(app.root, app.test));
    fs.mkdirSync(path.join('apps', options.name));
    fs.renameSync(path.join(app.root, app.tsconfig), 'tsconfig.app.json');
    fs.renameSync(path.join(app.root, app.testTsconfig), 'tsconfig.spec.json');
    fs.renameSync(path.join('e2e', 'tsconfig.e2e.json'), 'tsconfig.e2e.json');
    fs.renameSync(app.root, join('apps', options.name, app.root));
    fs.renameSync('e2e', join('apps', options.name, 'e2e'));

    return host;
  };
}

function dedup(array: any[]): any[] {
  const res = [];

  array.forEach(a => {
    if (res.indexOf(a) === -1) {
      res.push(a);
    }
  });
  return res;
}

export default function(options: Schema): Rule {
  return chain([
    moveFiles(options), branchAndMerge(chain([
      mergeWith(apply(url('./files'), [])),
    ])),
    updatePackageJson(), updateAngularCLIJson(options), updateTsConfigsJson(options), updateProtractorConf()
  ]);
}
