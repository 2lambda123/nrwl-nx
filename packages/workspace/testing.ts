export { MockBuilderContext } from './src/utils/testing-utils';
