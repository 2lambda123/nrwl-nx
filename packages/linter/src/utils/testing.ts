import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { join } from 'path';
import { Tree } from '@angular-devkit/schematics';

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/linter-migrations',
  join(__dirname, '../../migrations.json')
);

const testRunner = new SchematicTestRunner(
  '@nrwl/linter',
  join(__dirname, '../../collection.json')
);

export function runMigration<T = any>(
  migrationName: string,
  options: T,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync<T>(migrationName, options, tree)
    .toPromise();
}

export function runSchematic<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}
