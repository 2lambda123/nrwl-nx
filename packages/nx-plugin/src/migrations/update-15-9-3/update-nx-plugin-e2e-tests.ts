import {
  createProjectGraphAsync,
  formatFiles,
  readProjectConfiguration,
  Tree,
  visitNotIgnoredFiles,
} from '@nrwl/devkit';
import { JestExecutorOptions } from '@nrwl/jest/src/executors/jest/schema';
import { TEST_FILE_PATTERN } from '@nrwl/jest/src/utils/ast-utils';
import { forEachExecutorOptionsInGraph } from '@nrwl/devkit/src/generators/executor-options-utils';
import { tsquery } from '@phenomnomnominal/tsquery';
import { CallExpression, Node, StringLiteral, SyntaxKind } from 'typescript';

export async function updateNxPluginE2eTests(tree: Tree) {
  const graph = await createProjectGraphAsync();
  forEachExecutorOptionsInGraph<JestExecutorOptions>(
    graph,
    '@nrwl/nx-plugin:e2e',
    (options, projectName) => {
      const projectConfig = readProjectConfiguration(tree, projectName);

      visitNotIgnoredFiles(tree, projectConfig.root, (file) => {
        if (!TEST_FILE_PATTERN.test(file)) {
          return;
        }
        updateEnsureNxProject(tree, file);
      });
    }
  );
  await formatFiles(tree);
}

export function updateEnsureNxProject(tree: Tree, filePath: string) {
  let contents = tree.read(filePath, 'utf-8');

  if (!contents.includes('ensureNxProject')) {
    return;
  }

  const ensureNodes = tsquery.query(
    contents,
    ':matches(ImportDeclaration):has(Identifier[name="ensureNxProject"]):has(StringLiteral[value="@nrwl/devkit"])'
  );

  if (ensureNodes.length === 0) {
    return;
  }

  // Replace calls to ensureNxProject to use a library
  contents = tsquery.replace(
    contents,
    'CallExpression:has(Identifier[name="ensureNxProject"])',
    (node: CallExpression) => {
      const text = node.getText();
      if (node.arguments.length !== 2 || !text.startsWith('ensureNxProject(')) {
        return;
      }
      //TODO: resolve packages to libraries
      const npmPackage = (node.arguments[0] as StringLiteral).text;
      if (npmPackage !== '@lib/my-lib') {
        return;
      }

      if (node.parent.kind === SyntaxKind.AwaitExpression) {
        return `ensureNxProject('my-lib')`;
      }
      return `await ensureNxProject('my-lib')`;
    }
  );

  // The new signature for ensureNxProject is a promise that needs to be awaited
  // Update wrapping function to be async
  tsquery.query(contents, 'AwaitExpression').forEach((node: Node) => {
    let wrappingArrowFunction = node.parent;
    while (
      wrappingArrowFunction != null &&
      wrappingArrowFunction.kind !== SyntaxKind.ArrowFunction
    ) {
      wrappingArrowFunction = wrappingArrowFunction.parent;
    }

    if (wrappingArrowFunction == null) {
      return;
    }

    const text = wrappingArrowFunction.getText();
    if (!text.startsWith('async ()')) {
      contents = contents.replace(text, `async ${text}`);
    }
  });

  tree.write(filePath, contents);
}

export default updateNxPluginE2eTests;
