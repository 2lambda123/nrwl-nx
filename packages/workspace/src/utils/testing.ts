import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import type { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/express',
  join(__dirname, '../../../express/generators.json')
);

testRunner.registerCollection(
  '@nrwl/react',
  join(__dirname, '../../../react/generators.json')
);

testRunner.registerCollection(
  '@nrwl/angular',
  join(__dirname, '../../../angular/generators.json')
);

testRunner.registerCollection(
  '@nrwl/next',
  join(__dirname, '../../../next/generators.json')
);

testRunner.registerCollection(
  '@nrwl/node',
  join(__dirname, '../../../node/generators.json')
);

testRunner.registerCollection(
  '@nrwl/nest',
  join(__dirname, '../../../nest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/web',
  join(__dirname, '../../../web/generators.json')
);

/**
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nrwl/devkit. Generators can be tested by simply calling them in it().
 */
export function runSchematic<T extends object = any>(
  schematicName: string,
  options: T,
  tree: Tree
) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

/**
 * @deprecated This will be removed in v17. Prefer writing Nx Generators with @nrwl/devkit. Generators can be tested by simply calling them in it().
 */
export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
