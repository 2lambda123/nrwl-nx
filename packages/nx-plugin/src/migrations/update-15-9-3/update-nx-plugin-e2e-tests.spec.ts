import {
  logger,
  ProjectGraph,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { libraryGenerator } from '@nrwl/workspace';
import { updateNxPluginE2eTests } from './update-nx-plugin-e2e-tests';

let projectGraph: ProjectGraph;

jest.mock('@nrwl/devkit', () => ({
  ...jest.requireActual<any>('@nrwl/devkit'),
  createProjectGraphAsync: jest
    .fn()
    .mockImplementation(async () => projectGraph),
}));

describe('Nx Plugin Migration - ensureNxProject usage in e2e tests', () => {
  let tree: Tree;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it('should not update anything if test does not have references to ensureNxProject', async () => {
    await setup(tree, 'my-lib');

    const expected = tree.read('libs/my-lib/tests/empty-test.spec.ts', 'utf-8');
    await updateNxPluginE2eTests(tree);

    expect(tree.read('libs/my-lib/tests/empty-test.spec.ts', 'utf-8')).toEqual(
      expected
    );
  });

  it('should update calls to ensureNxProject and use project instead of package name', async () => {
    await setup(tree, 'my-lib');

    await updateNxPluginE2eTests(tree);

    expect(
      tree.read('libs/my-lib/tests/test.spec.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should be idempotent', async () => {
    await setup(tree, 'my-lib');

    await updateNxPluginE2eTests(tree);
    await updateNxPluginE2eTests(tree);

    expect(
      tree.read('libs/my-lib/tests/test.spec.ts', 'utf-8')
    ).toMatchSnapshot();
  });
});

async function setup(tree: Tree, name: string) {
  await libraryGenerator(tree, { name });
  const projectConfig = readProjectConfiguration(tree, name);
  projectConfig.targets['e2e'] = {
    executor: '@nrwl/nx-plugin:e2e',
    options: {
      jestConfig: `libs/${name}/jest.config.ts`,
    },
  };
  updateProjectConfiguration(tree, name, projectConfig);

  tree.write(
    `libs/${name}/tests/test.spec.ts`,
    `import { runNxCommandAsync, uniq } from '@nrwl/nx-plugin/testing';

import { ensureNxProject } from '@nrwl/devkit';

describe('${name} e2e', () => {
  // Setting up individual workspaces per
  // test can cause e2e runs to take a long time.
  // For this reason, we recommend each suite only
  // consumes 1 workspace. The tests should each operate
  // on a unique project in the workspace, such that they
  // are not dependant on one another.
  beforeAll(() => {
    ensureNxProject('@lib/${name}', 'dist/packages/${name}');
  });

  beforeAll(async () => {
    await ensureNxProject('@lib/${name}', 'dist/packages/${name}');
  });

  afterAll(() => {
    // nx reset kills the daemon, and performs
    // some work which can help clean up e2e leftovers
    runNxCommandAsync('reset');
  });
});
`
  );

  tree.write(
    `libs/${name}/tests/empty-test.spec.ts`,
    `import { runNxCommandAsync, uniq } from '@nrwl/nx-plugin/testing';

describe('${name} e2e', () => {
  afterAll(() => {
    // nx reset kills the daemon, and performs
    // some work which can help clean up e2e leftovers
    runNxCommandAsync('reset');
  });
});
`
  );

  projectGraph = {
    dependencies: {},
    nodes: {
      [name]: {
        name,
        type: 'lib',
        data: projectConfig,
      } as any,
    },
  };
}
