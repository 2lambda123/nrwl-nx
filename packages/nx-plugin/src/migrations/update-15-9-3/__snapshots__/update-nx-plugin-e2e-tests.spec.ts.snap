// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nx Plugin Migration - ensureNxProject usage in e2e tests should be idempotent 1`] = `
"import { runNxCommandAsync, uniq } from '@nrwl/nx-plugin/testing';

import { ensureNxProject } from '@nrwl/devkit';

describe('my-lib e2e', () => {
  // Setting up individual workspaces per
  // test can cause e2e runs to take a long time.
  // For this reason, we recommend each suite only
  // consumes 1 workspace. The tests should each operate
  // on a unique project in the workspace, such that they
  // are not dependant on one another.
  beforeAll(async () => {
    await ensureNxProject('my-lib');
  });

  beforeAll(async () => {
    await ensureNxProject('my-lib');
  });

  afterAll(() => {
    // nx reset kills the daemon, and performs
    // some work which can help clean up e2e leftovers
    runNxCommandAsync('reset');
  });
});
"
`;

exports[`Nx Plugin Migration - ensureNxProject usage in e2e tests should update calls to ensureNxProject and use project instead of package name 1`] = `
"import { runNxCommandAsync, uniq } from '@nrwl/nx-plugin/testing';

import { ensureNxProject } from '@nrwl/devkit';

describe('my-lib e2e', () => {
  // Setting up individual workspaces per
  // test can cause e2e runs to take a long time.
  // For this reason, we recommend each suite only
  // consumes 1 workspace. The tests should each operate
  // on a unique project in the workspace, such that they
  // are not dependant on one another.
  beforeAll(async () => {
    await ensureNxProject('my-lib');
  });

  beforeAll(async () => {
    await ensureNxProject('my-lib');
  });

  afterAll(() => {
    // nx reset kills the daemon, and performs
    // some work which can help clean up e2e leftovers
    runNxCommandAsync('reset');
  });
});
"
`;
